auth/api/iam/s2s_vptoken.go

```diff
@@ -40,40 +41,106 @@ import (
+		return nil, oauth.OAuth2Error{
+			Code:          oauth.InvalidRequest,
+			Description:   "assertion parameter is invalid: invalid query escaping",
+			InternalError: err,
+		}
+	}
+	pexEnvelope, err := pe.ParseEnvelope([]byte(assertionDecoded))
+	if err != nil {
+		return nil, oauth.OAuth2Error{
+			Code:        oauth.InvalidRequest,
+			Description: "assertion parameter is invalid: " + err.Error(),
+		}
+	}
 
-func (s serviceToService) validateVPToken(params map[string]string) (string, error) {
-	submission := params["presentation_submission"]
-	scope := params["scope"]
-	vp_token := params["vp_token"]
-	if submission == "" || scope == "" || vp_token == "" {
-		// TODO: right error response
-		return "", errors.New("missing required parameters")
-	}
-	// TODO: https://github.com/nuts-foundation/nuts-node/issues/2418
-	// TODO: verify parameters
-	return scope, nil
-}
+	// Unmarshal presentation submission
+	submissionDecoded, err := url.QueryUnescape(submissionEncoded)
+	if err != nil {
+		return nil, oauth.OAuth2Error{
+			Code:          oauth.InvalidRequest,
+			Description:   "presentation_submission parameter is invalid: invalid query escaping",
+			InternalError: err,
+		}
+	}
+	submission, err := pe.ParsePresentationSubmission([]byte(submissionDecoded))
+	if err = json.Unmarshal([]byte(submissionDecoded), &submission); err != nil {
+		return nil, oauth.OAuth2Error{
+			Code:          oauth.InvalidRequest,
+			Description:   "presentation_submission parameter is invalid: invalid JSON",
+			InternalError: err,
+		}
+	}
+
+	for _, presentation := range pexEnvelope.Presentations {
+		err = credential.VerifyPresenterIsCredentialSubject(presentation)
+		if err != nil {
+			return nil, oauth.OAuth2Error{
+				Code:        oauth.InvalidRequest,
+				Description: fmt.Sprintf("verifiable presentation is invalid: %s", err.Error()),
+			}
+		}
+	}
+
+	// Validate the presentation submission:
+	// 1. Resolve presentation definition for the requested scope
+	// 2. Check submission against presentation and definition
+	definition := s.auth.PresentationDefinitions().ByScope(scope)
+	if definition == nil {
+		return nil, oauth.OAuth2Error{
+			Code:        oauth.InvalidScope,
+			Description: fmt.Sprintf("unsupported scope for presentation exchange: %s", scope),
+		}
+	}
+
+	_, err = submission.Validate(*pexEnvelope, *definition)
+	if err != nil {
+		return nil, oauth.OAuth2Error{
+			Code:          oauth.InvalidRequest,
+			Description:   "presentation submission does not conform to Presentation Definition",
+			InternalError: err,
+		}
+	}
 
-func (s serviceToService) handleAuthzRequest(_ map[string]string, _ *Session) (*authzResponse, error) {
-	// Protocol does not support authorization code flow
-	return nil, nil
+	// Check signatures of VP and VCs. Trust should be established by the Presentation Definition.
+	for _, presentation := range pexEnvelope.Presentations {
+		_, err = s.vcr.Verifier().VerifyVP(presentation, true, true, nil)
```

Comment:
`VerifyVP(..., validAt)` should not be nil?

Reply:
Defaults to current time